### DISCUSSION TITLE
Capturing USB camera images using opencvsharp is not smooth and stuck

### DISCUSSION THREAD

**Discussion by mimimiaomiao1** [AvaloniaUI/Avalonia#11074] [General]
There's no problem with wpf

**Comment by timunie** [comment]
Please try using latest nightly as rendering performance was improved. Maybe this helps. Otherwise, we need to know where the bottleneck explicitly is.

**Reply by mimimiaomiao1** [reply]
> Please try using latest nightly as rendering performance was improved. Maybe this helps. Otherwise, we need to know where the bottleneck explicitly is.

Do you have any examples? I don't know how to implement it

**Reply by timunie** [reply]
Me neither. Acc. to the issue title I thought you have a way but it is not fast enough. If not, I have no idea. Probably a WriteableBitmap can be used to render the camera image, but I'm just guessing. Maybe check Render-demo in source code: https://github.com/AvaloniaUI/Avalonia/blob/master/samples/RenderDemo/Pages/RenderTargetBitmapPage.cs

**Comment by Tryanks** [comment]
Hi, I'm sorry my statement can't be helpful on this issue.

Can you share exactly how you used OpenCVSharpb to capture the USB camera and render it in Avalonia?

I tried using Bitmap and data binding, but this caused a lot of GC and memory waste (the discarded frames remained in memory waiting to be cleaned up).

**Reply by mimimiaomiao1** [reply]
> Hi, I'm sorry my statement can't be helpful on this issue.
> 
> Can you share exactly how you used OpenCVSharpb to capture the USB camera and render it in Avalonia?
> 
> I tried using Bitmap and data binding, but this caused a lot of GC and memory waste (the discarded frames remained in memory waiting to be cleaned up).

I didn't use binding. Opencv only returned bitmap

(pseudo code)
Avalonia image control source=opencv.bitmap

**Reply by mimimiaomiao1** [reply]
> Hi, I'm sorry my statement can't be helpful on this issue.
> 
> Can you share exactly how you used OpenCVSharpb to capture the USB camera and render it in Avalonia?
> 
> I tried using Bitmap and data binding, but this caused a lot of GC and memory waste (the discarded frames remained in memory waiting to be cleaned up).

咱们可以讲中文.

**Reply by kmate95** [reply]
This is a working (as of preview6) example I've been using with OpenCVSharp. Has some parts specific to my use case (that is similar to OP but with custom hardware) but most of it is general. Sharing this because it was VERY hard to find this solution. (as a newcomer to Avalonia)
```csharp
/// <summary>
/// Converts a OpenCV Mat to a bitmap.
/// </summary>
public class MatBitmapValueConverter : IValueConverter {
  public static readonly MatBitmapValueConverter Instance = new MatBitmapValueConverter();

  public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture) {
    if (value == null)
      return null;
    // if (!targetType.IsAssignableFrom(typeof(IImage)))
    //   return null;
    if (!targetType.IsAssignableFrom(typeof(WriteableBitmap)))
      return new BindingNotification(new NotSupportedException(), BindingErrorType.Error);

    var mat = value switch {
//      MatImageMessage msg => msg.Image,
      Mat mata            => mata,
      _                   => null,
    };
    if (mat == null) return new BindingNotification(new NotSupportedException(), BindingErrorType.Error);
    if (parameter is WriteableBitmap wb) {
      try { // may not be good size
        mat.ToBitmapParallel(wb);
        return wb;
      }
      catch (ArgumentException e) {
        // ignored
//        App.TryGetLogger<MatBitmapValueConverter>()?.LogError(e, "Error converting to bitmap");
      }
    }
    return mat.ToBitmapParallel(); // this just creates a new WriteableBitmap and calls method defined below
  }

  public object ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture) {
    return new BindingNotification(new NotSupportedException(), BindingErrorType.Error);
  }
}
```

```csharp
/// <summary>
///     Converts Mat to WriteableBitmap.
///     This method is more efficient because new instance of WriteableBitmap is not allocated.
///     Original:
///     https://github.com/shimat/opencvsharp/blob/master/src/OpenCvSharp.WpfExtensions/WriteableBitmapConverter.cs
/// </summary>
/// <param name="mat">Input Mat</param>
/// <param name="dst">Output WriteableBitmap</param>
[SuppressMessage("ReSharper", "AccessToDisposedClosure")]
public static void ToBitmapParallel(this Mat mat, WriteableBitmap dst) {
  ArgumentNullException.ThrowIfNull(mat);
  ArgumentNullException.ThrowIfNull(dst);

  if (mat.Width != dst.PixelSize.Width || mat.Height != dst.PixelSize.Height)
    throw new ArgumentException("size of src must be equal to size of dst");
  if (dst.Format != PixelFormat.Bgra8888 || !dst.Dpi.Equals(new Vector(96, 96)))
    throw new ArgumentException("Currently only Bgra8888 + 96 DPI WriteableBitmaps can be reused ");


  var width = mat.Width;
  var height = mat.Height;

  using var lockBuffer = dst.Lock();
  var stride = lockBuffer.RowBytes;
  var bufferAddress = lockBuffer.Address;
  unsafe {
    switch (mat.Channels()) {
      // Method 1 (Span copy)
      case 1:
        Parallel.For(0, height, y => {
          var spanBitmap = new Span<uint>(IntPtr.Add(bufferAddress, y * stride).ToPointer(), width); // lockBuffer.GetPixelRowSpan(y,width);
          var spanMat = mat.GetRowSpan<byte>(y, width);
          for (var x = 0; x < width; x++) {
            var color = spanMat[x];
            spanBitmap[x] = (uint) (color | (color << 8) | (color << 16) | (0xff << 24));
          }
        });
        break;
      case 3:
        Parallel.For(0, height, y => {
          var spanBitmap = new Span<uint>(IntPtr.Add(bufferAddress, y * stride).ToPointer(), width); // lockBuffer.GetPixelRowSpan(y,width);
          var spanMat = mat.GetRowSpan<byte>(y);
          var pixel = 0;
          for (var x = 0; x < width; x++) spanBitmap[x] = (uint) (spanMat[pixel++] | (spanMat[pixel++] << 8) | (spanMat[pixel++] << 16) | (0xff << 24));
        });

        break;
      case 4:
        if (mat.Type() == MatType.CV_8U) {
          Parallel.For(0, height, y => {
            var spanBitmap = new Span<uint>(IntPtr.Add(bufferAddress, y * stride).ToPointer(), width); // lockBuffer.GetPixelRowSpan(y,width);
            var spanMat = mat.GetRowSpan<byte>(y);
            var pixel = 0;
            for (var x = 0; x < width; x++) spanBitmap[x] = (uint) (spanMat[pixel++] | (spanMat[pixel++] << 8) | (spanMat[pixel++] << 16) | (spanMat[pixel++] << 24));
          });
        }
        else if (mat.Type() == MatType.CV_32S) {
          var dataCount = width * height;
          mat.GetDataSpan<uint>().CopyTo(new Span<uint>(lockBuffer.Address.ToPointer(), dataCount));
        }

        break;
    }
  }
}
public static unsafe Span<T> GetRowSpan<T>(this Mat mat, int y, int length = 0, int offset = 0)
  => new(IntPtr.Add(mat.DataStart, y * mat.GetRealStep() + offset).ToPointer(), length <= 0 ? mat.GetRealStep() : length);
public static int GetRealStep(this Mat mat) 
  =>  mat.Width * mat.Channels();
```
```xaml
<Image Name="ScanImage"
             HorizontalAlignment="Stretch" VerticalAlignment="Stretch"
             MaxWidth="{Binding $parent.Bounds.Width}"
             MaxHeight="{Binding $parent.Bounds.Height}"
             Source="{ Binding ScannerViewModel.CurrentScanImage,
             Converter={x:Static avalonia:MatBitmapValueConverter.Instance},
             ConverterParameter={Binding $self.Source}}" />
```

**Comment by yangjieshao** [comment]
https://github.com/kekyo/FlashCap

用 [FlashCap](https://github.com/kekyo/FlashCap)  跨平台获取帧数据
例子里用到了  [SkiaImageView](https://github.com/kekyo/SkiaImageView) 进行渲染，是同一个作者写的
性能的话  反正信创用的arm的电脑 60帧没问题


只是获取usb摄像头画面的话 opencvsharp  太重了
你要是只考虑Windows的话  可以考虑 [UsbCamera](https://github.com/secile/UsbCamera) 获取帧数据

**Reply by mimimiaomiao1** [reply]
> https://github.com/kekyo/FlashCap
> 
> 用 [FlashCap](https://github.com/kekyo/FlashCap) 跨平台获取帧数据 例子里用到了 [SkiaImageView](https://github.com/kekyo/SkiaImageView) 进行渲染，是同一个作者写的 性能的话 反正信创用的arm的电脑 60帧没问题
> 
> 只是获取usb摄像头画面的话 opencvsharp 太重了 你要是只考虑Windows的话 可以考虑 [UsbCamera](https://github.com/secile/UsbCamera) 获取帧数据

我用opencv是因为还有一些视觉方面相关的处理.
您说的这个支持设置分辨率,设置解码格式吗?

**Reply by yangjieshao** [reply]
> > https://github.com/kekyo/FlashCap
> > 用 [FlashCap](https://github.com/kekyo/FlashCap) 跨平台获取帧数据 例子里用到了 [SkiaImageView](https://github.com/kekyo/SkiaImageView) 进行渲染，是同一个作者写的 性能的话 反正信创用的arm的电脑 60帧没问题
> > 只是获取usb摄像头画面的话 opencvsharp 太重了 你要是只考虑Windows的话 可以考虑 [UsbCamera](https://github.com/secile/UsbCamera) 获取帧数据
> 
> 我用opencv是因为还有一些视觉方面相关的处理. 您说的这个支持设置分辨率,设置解码格式吗?


可以通过 `devices.EnumerateDescriptors()` 获取设备支持的流的分辨率和格式(JPEG/YUYV) 然后直接从设备获取对应的流

**Reply by mimimiaomiao1** [reply]
> > > https://github.com/kekyo/FlashCap
> > > 用 [FlashCap](https://github.com/kekyo/FlashCap) 跨平台获取帧数据 例子里用到了 [SkiaImageView](https://github.com/kekyo/SkiaImageView) 进行渲染，是同一个作者写的 性能的话 反正信创用的arm的电脑 60帧没问题
> > > 只是获取usb摄像头画面的话 opencvsharp 太重了 你要是只考虑Windows的话 可以考虑 [UsbCamera](https://github.com/secile/UsbCamera) 获取帧数据
> > 
> > 
> > 我用opencv是因为还有一些视觉方面相关的处理. 您说的这个支持设置分辨率,设置解码格式吗?
> 
> 可以通过 `devices.EnumerateDescriptors()` 获取设备支持的流的分辨率和格式(JPEG/YUYV) 然后直接从设备获取对应的流

还有一个疑问,我并不是使用的mvvm方式.可以使用SkiaImageView那个吗?  还有可不可以在图像上画直线,画矩形. Line Rectangle

**Reply by yangjieshao** [reply]
> 还有一个疑问,我并不是使用的mvvm方式.可以使用SkiaImageView那个吗? 还有可不可以在图像上画直线,画矩形. Line Rectangle

你可以看一下 sample 的代码
https://github.com/kekyo/FlashCap/tree/main/samples/FlashCap.Avalonia/FlashCap.Avalonia.UI
```
<siv:SKImageView
            Source="{Binding Image}"
            Stretch="Uniform" />
```
你直接更新`Source`就是了  
给到 `Source` 的是 `SKBitmap ` 你可以先在 `SKBitmap` 画东西  也可以在 `<siv:SKImageView/>`  上覆盖一层透明的Grid 随便画啥都可以了就

**Reply by yangjieshao** [reply]
> 还有一个疑问,我并不是使用的mvvm方式.可以使用SkiaImageView那个吗? 还有可不可以在图像上画直线,画矩形. Line Rectangle


对了 原版的支持到 0.10.0
你可以用我给升级到 11.0.0-preview6 的
https://github.com/yangjieshao/SkiaImageView/tree/main/SkiaImageView.Avalonia

**Reply by mimimiaomiao1** [reply]
> > 还有一个疑问,我并不是使用的mvvm方式.可以使用SkiaImageView那个吗? 还有可不可以在图像上画直线,画矩形. Line Rectangle
> 
> 对了 原版的支持到 0.10.0 你可以用我给升级到 11.0.0-preview6 的 https://github.com/yangjieshao/SkiaImageView/tree/main/SkiaImageView.Avalonia

我应该怎么使用这个预览版的...是在Nuget升级吗?还是如何操作

**Reply by yangjieshao** [reply]
> 我应该怎么使用这个预览版的...是在Nuget升级吗?还是如何操作

下载 编译 打包

**Comment by qq908382818** [comment]
我用OPENCVSHARP可以正常拿到图片，但是没办法显示到AVA的IMAGE控件中。怎么转换MAT到IMAGE控件中显示呀，求救，不要介绍其他组件。谢谢；[#15234](https://github.com/AvaloniaUI/Avalonia/discussions/15234)

**Comment by Pratham95VP** [comment]
Sorry but are you facing issue while displaying video feed from usb camera or capturing frame/Image ? If you are facing issue as video feed from usb camera is not smooth then `VideoCapture(int index, VideoCaptureAPIs apiPreference = VideoCaptureAPIs.ANY)` try changing apiPreference. If you are facing issue while capturing image and display it into Avalonia.Controls.Image then you need to use WriteableBitmap class.
